plugin/plugin.go

```diff
@@ -1,243 +1,164 @@
-type CriticalAttributes struct {
-	ContentType          string                      `json:"contentType"`
-	SigningScheme        string                      `json:"signingScheme"`
-	Expiry               *time.Time                  `json:"expiry,omitempty"`
-	AuthenticSigningTime *time.Time                  `json:"authenticSigningTime,omitempty"`
-	ExtendedAttributes   map[interface{}]interface{} `json:"extendedAttributes,omitempty"`
-}
+func run(ctx context.Context, pluginName string, pluginPath string, cmder commander, req proto.Request, resp interface{}) error {
+	// serialize request
+	data, err := json.Marshal(req)
+	if err != nil {
+		return pluginErr(pluginName, fmt.Errorf("failed to marshal request object: %w", err))
+	}
 
-// TrustPolicy represents trusted identities that sign the artifacts
-type TrustPolicy struct {
-	TrustedIdentities     []string                 `json:"trustedIdentities"`
-	SignatureVerification []VerificationCapability `json:"signatureVerification"`
-}
+	// execute request
+	out, err := cmder.Output(ctx, pluginPath, string(req.Command()), data)
+	if err != nil {
+		var re proto.RequestError
+		err = json.Unmarshal(out, &re)
+		if err != nil {
+			return proto.RequestError{Code: proto.ErrorCodeGeneric, Err: fmt.Errorf("error: %v. stderr: %s", err, out)}
+		}
+		return re
+	}
 
-func (VerifySignatureRequest) Command() Command {
-	return CommandVerifySignature
+	// deserialize response
+	err = json.Unmarshal(out, resp)
+	if err != nil {
+		return fmt.Errorf("failed to decode json response: %w", ErrNotCompliant)
+	}
+	return nil
 }
 
-// VerifySignatureResponse is the response of a verify-signature request.
-type VerifySignatureResponse struct {
-	VerificationResults map[VerificationCapability]*VerificationResult `json:"verificationResults"`
-	ProcessedAttributes []interface{}                                  `json:"processedAttributes"`
+// commander is defined for mocking purposes.
+type commander interface {
+	// Output runs the command, passing req to the its stdin.
+	// It only returns an error if the binary can't be executed.
+	// Returns stdout if err is nil, stderr if err is not nil.
+	Output(ctx context.Context, path string, command string, req []byte) (out []byte, err error)
 }
 
-// VerificationResult is the result of a verification performed by the plugin
-type VerificationResult struct {
-	Success bool   `json:"success"`
-	Reason  string `json:"reason,omitempty"`
-}
+// execCommander implements the commander interface using exec.Command().
+type execCommander struct{}
 
-// Request defines a plugin request, which is always associated to a command.
-type Request interface {
-	Command() Command
+func (c execCommander) Output(ctx context.Context, name string, command string, req []byte) ([]byte, error) {
+	var stdout, stderr bytes.Buffer
+	cmd := exec.CommandContext(ctx, name, command)
+	cmd.Stdin = bytes.NewReader(req)
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err := cmd.Run()
```

Comment:
Will [Cmd.Output](https://pkg.go.dev/os/exec#Cmd.Output) be a better implementation?

Reply:
Updated.